tools/extract_features_from_lyric.py

```diff
@@ -0,0 +1,310 @@
+                if i+1 < len(doc):
+                    i += 1
+                    following_token = doc[i]
+                    if following_token is not None and following_token.tag_ == 'VB':
+                        verb_tense_freq['future'] += 1
+
+    return verb_tense_freq['present'], verb_tense_freq['past'], verb_tense_freq['future']
+
+
+def get_pos_tags_count(lines, nlp):
+    pos_tags_count = {
+            'ADJ': 0,                   #adjective
+            'ADP': 0,                   #adposition
+            'ADV': 0,                   #adverb
+            'AUX': 0,                   #auxiliary
+            'CCONJ': 0,                 #coordinating conjunction
+            'DET': 0,                   #determiner
+            'INTJ': 0,                  #interjection
+            'NOUN': 0,                  #noun
+            'NUM': 0,                   #numeral
+            'PART': 0,                  #particle
+            'PRON': 0,                  #pronoun
+            'PROPN': 0,                 #proper noun
+            'PUNCT': 0,                 #punctuation
+            'SCONJ': 0,                 #subordinating conjunction
+            'SYM': 0,                   #symbol
+            'VERB': 0,                  #verb
+            'X': 0,                     #other
+        }
+
+    for line in lines:
+        doc = nlp(line)
+        for token in doc:
+            if token.pos_ in pos_tags_count.keys():
+                pos_tags_count[token.pos_] += 1
+    
+    return pos_tags_count
+
+
+def get_sentiment(lyric):
+    polarity = textblob.TextBlob(lyric).sentiment.polarity
+    subjectivity = textblob.TextBlob(lyric).sentiment.subjectivity
+    return polarity, subjectivity
+
+
+def extract_all_features(df):
+
+    nlp = spacy.load('en_core_web_lg')
+    nlp.vocab.add_flag(lambda s: s.lower() in spacy.lang.en.stop_words.STOP_WORDS, spacy.attrs.IS_STOP)
+
+    rows = list()
+    ids = list()
+    for index, row in df.iterrows():
+        
+        title = row['title']
+        lyric, lines = clean_lyric(row['lyric'], title)
+        
+        #features
+        lyrics_vector = get_lyrics_vector(lyric, nlp)
+        echoisms = get_echoisms(lines, nlp)
+        duplicate_lines = get_duplicate_lines(lines)
+        title_in_lyric = is_title_in_lyric(title, lines)
+        verb_present_freq, verb_past_freq, verb_future_freq = get_verb_tense_freq(lyric, nlp)
+        pos_tags_count = get_pos_tags_count(lines, nlp)
+        sentiment_polarity, sentiment_subjectivity = get_sentiment(lyric)
+
+        row = [lyrics_vector, echoisms, duplicate_lines, title_in_lyric, verb_present_freq, verb_past_freq, verb_future_freq,
+                pos_tags_count['ADJ'], pos_tags_count['PUNCT'], sentiment_polarity, sentiment_subjectivity]
+
+        rows.append(row)
+        ids.append(index)
+
+        #In the end, we obtained the best results just by using the following features: 
```

Comment:
This type of redundant code (long comments) should be avoided where it's possible - e.g., you have variable names like that, why do u include that in the comment as well? If it comes to echoism, lyrics vector, sentiment, subjectivity degree you can put it near the variable initialization i.e. for example, above echoism=get_echoism(...,...) line

Reply:
Right!